src/ic-cdk/src/api/call.rs

```diff
@@ -189,6 +206,258 @@ unsafe extern "C" fn cleanup<T: AsRef<[u8]>>(state_ptr: *const RwLock<CallFuture
+    }
+}
+
+/// Methods to configure a call.
+pub trait ConfigurableCall {
+    /// Sets the cycles payment for the call.
+    ///
+    /// If invoked multiple times, the last value is used.
+    fn with_cycles(self, cycles: u128) -> Self;
+
+    /// Sets the call to have a guaranteed response.
+    ///
+    /// If [change_timeout](ConfigurableCall::change_timeout) is invoked after this method,
+    /// the call will instead be set with Best-Effort Responses.
+    fn with_guaranteed_response(self) -> Self;
+
+    /// Sets the timeout for the Best-Effort Responses.
+    ///
+    /// If not set, the call will default to a 10 seconds timeout.
+    /// If invoked multiple times, the last value is used.
+    /// If [with_guaranteed_response](ConfigurableCall::with_guaranteed_response) is invoked after this method,
+    /// the timeout will be ignored.
+    fn change_timeout(self, timeout_seconds: u32) -> Self;
+}
+
+impl<'a> ConfigurableCall for Call<'a> {
+    fn with_cycles(mut self, cycles: u128) -> Self {
+        self.payment = Some(cycles);
+        self
+    }
+
+    fn with_guaranteed_response(mut self) -> Self {
+        self.timeout_seconds = None;
+        self
+    }
+
+    fn change_timeout(mut self, timeout_seconds: u32) -> Self {
+        self.timeout_seconds = Some(timeout_seconds);
+        self
+    }
+}
+
+impl<'a, T> ConfigurableCall for CallWithArgs<'a, T> {
+    fn with_cycles(mut self, cycles: u128) -> Self {
+        self.call.payment = Some(cycles);
+        self
+    }
+
+    fn with_guaranteed_response(mut self) -> Self {
+        self.call.timeout_seconds = None;
+        self
+    }
+
+    fn change_timeout(mut self, timeout_seconds: u32) -> Self {
+        self.call.timeout_seconds = Some(timeout_seconds);
+        self
+    }
+}
+
+impl<'a, A> ConfigurableCall for CallWithRawArgs<'a, A> {
+    fn with_cycles(mut self, cycles: u128) -> Self {
+        self.call.payment = Some(cycles);
+        self
+    }
+
+    fn with_guaranteed_response(mut self) -> Self {
+        self.call.timeout_seconds = None;
+        self
+    }
+
+    fn change_timeout(mut self, timeout_seconds: u32) -> Self {
+        self.call.timeout_seconds = Some(timeout_seconds);
+        self
+    }
+}
+
+/// Methods to send a call.
+pub trait SendableCall {
+    /// Sends the call and gets the reply as raw bytes.
+    fn call_raw(self) -> impl Future<Output = CallResult<Vec<u8>>> + Send + Sync;
+
+    /// Sends the call and decodes the reply to a Candid type.
+    fn call<R: for<'b> ArgumentDecoder<'b>>(
+        self,
+    ) -> impl Future<Output = CallResult<R>> + Send + Sync
+    where
+        Self: Sized,
+    {
+        let fut = self.call_raw();
+        async {
+            let bytes = fut.await?;
+            decode_args(&bytes).map_err(decoder_error_to_reject::<R>)
+        }
+    }
+
+    /// Sends the call and decodes the reply to a Candid type with a decoding quota.
+    fn call_with_decoder_config<R: for<'b> ArgumentDecoder<'b>>(
```

Comment:
With best-effort call, we can merge this function with the original `call` function. If the call is best-effort, put a decoding quota, if it's not, decode without the quota.

Reply:
The "decoding quota" and "best effort response" are orthogonal concepts. And they affect different stages of the inter-canister call. So I insist on keeping them separate.